ocaml/typing/typedtree.mli

```diff
@@ -296,10 +296,10 @@ and expression_desc =
                          (Labelled "y", Some (Texp_constant Const_int 3))
                         ])
 
-            The [Zero_alloc_utils.Assume_info.t] records the optional
-            [@zero_alloc assume] attribute that may appear on applications.  If
-            that attribute is absent, it is [Assume_info.none].
-          *)
+            The [zero_alloc_attribute] records the optional [@zero_alloc assume]
+            attribute that may appear on applications.  Invariant: it is either
+            [Default_zero_alloc] (if the user wrote no attribute) or [Assume]
+            (if they did). *)
```

Comment:
Why not mint a new type and remove the failure cases? (I'm currently confused how `Ignore_assert_all` is ruled out by type-checking, and I think minting a new type here makes that easier to see/understand.)

Reply:
I've taken a stab at this. It didn't result in any major simplfications, but is probably a bit clearer - let me know what you think.